Implement Swift support for Objective-C resilient class stubs

slavapestov opened the pull request:
When a class has resilient ancestry, the metadata has to be dynamically allocated because its size is not known at compile time. This means Clang has nothing to reference if the class is exposed to Objective-C.

Class stubs solve this problem. Swift emits a class stub for the class, storing a reference to a callback function which realizes the actual class metadata at runtime. The class stub is exported via the usual `OBJC_CLASS_$_Foo` symbol. In the generated header, we print a new `objc_class_stub` attribute for classes with resilient ancestry.

When Clang references a class with a class stub, the emitted classref has the least significant bit of the class object pointer set to 1. Instead of loading directly from the classref, Clang emits a call to `objc_loadClassref()`. This function checks if the least significant bit of the class pointer at the given class ref is 1; if it is, it calls the callback function to produce the actual class metadata, which is then stored back into the classref for future accesses.

Nothing much changes on the Swift side; we still use class metadata accesses as usual. The one case where Swift references a class stub is when emitting a category for a class with resilient ancestry. This case is not supported today, without class stubs.

In this PR, the whole enchilada is guarded by a new `-enable-objc-resilient-class-stubs` frontend flag. When this flag is off, there should be no change in behavior, except I added some diagnostics for the unsupported case of defining a category, whereas previously we would crash with an IRGen assertion.

rdar://problem/45647064

slavapestov committed:
AST: Add ModuleDecl::isResilient()

This cleans up some code. I'm keeping the ResilienceStrategy enum around
though, in case we want to use it to version the ABI in the future.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
The corresponding Clang change is https://github.com/apple/swift-clang/pull/278

slavapestov committed:
AST: Generalize ClassDecl::checkObjCAncestry() to ClassDecl::checkAncestry()

This replaces ClassDecl::hasObjCMembers() and some hand-coded walks.

slavapestov committed:
PrintAsObjC: Add support for classes with resilient ancestry

Without -enable-resilient-objc-class-stubs, don't print classes with
resilient ancestry at all, since they're not actually visible to
Clang and referencing one will not actually work.

When the flag is specified, such classes do appear in the generated
header with the new objc_class_stub Clang attribute.

slavapestov committed:
Sema: Diagnose unsupported '@objc' on classes and members of extensions of classes with resilient ancestry

Unless -enable-resilient-objc-class-stubs is passed in, these cases
are not supported, so now we diagnose them instead of asserting or
failing to link.

Note the behavior change here; classes with resilient ancestry were
previously isObjC(). However this is wrong since isObjC() means
"statically visible to Objective-C via the generated header".

After this patch, isObjC() only returns true for a class with resilient
ancestry if -enable-resilient-objc-class-stubs is passed in.

slavapestov commented:
@jrose-apple @rjmccall This is ready for review, please take a look.

rjmccall commented on include/swift/AST/Decl.h:
```diff
@@ -554,12 +554,8 @@ class alignas(1 << DeclAlignInBits) Decl {
-    /// Whether this class has @objc members.
-    HasObjCMembersComputed : 1,
-    HasObjCMembers : 1,
+    /// Information about the class's ancestry.
```
Please refer to `AncestryOptions` in this comment.

rjmccall commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3594,9 +3594,11 @@ ERROR(invalid_nonobjc_extension,none,
+      "'@objc' members", ())
+ERROR(objc_in_resilient_extension,none,
+      "extensions of classes built with library evolution support "
+      "cannot contain '@objc' members", ())
```
Hmm.  Do we not have an ability to find and register these when the class is constructed?  I guess it would at least need runtime support and so be deployment-gated.

slavapestov replied:
This is what the PR is all about :) The new diagnostic is only emitted if LangOpts.EnableObjCResilientClassStubs is not set. In the future the flag will become a deployment target check and the diagnostic will be rephrased.

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -590,20 +590,22 @@ swift_relocateClassMetadata(const ClassDescriptor *descriptor,
+/// parameter.
+///
+/// Note: Swift 5.0 did not have an 'objcStub' parameter, so its value
+/// is not defined unless the flag is set.
```
Is this the only reasonable way to do this?  It can't be linked from the type descriptor or something?

slavapestov replied:
It could be, yeah. I'll change it if you'd prefer to avoid a proliferation of new parameters to this function.

rjmccall replied:
It's code size for the caller and a more complex CC rule, so if it can be safely added to the TD, that would be better.  Will the stub reference only exist when deploying to newer runtimes?

slavapestov replied:
> Will the stub reference only exist when deploying to newer runtimes?

Correct.

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -2959,7 +2959,7 @@ IRGenModule::getAddrOfMetaclassObject(ClassDecl *decl,
 llvm::Function *
 IRGenModule::getAddrOfObjCMetadataUpdateFunction(ClassDecl *classDecl,
                                                  ForDefinition_t forDefinition) {
-  IRGen.noteUseOfTypeMetadata(classDecl);
```
Is this not necessary?

slavapestov replied:
I don't think so, no. While it doesn't hurt, it's confusing to see it there when the only time we reference the update callback is while we're _already_ emitting the metadata.

rjmccall commented on lib/IRGen/GenHeap.cpp:
```diff
@@ -1900,7 +1892,7 @@ IsaEncoding irgen::getIsaEncodingForType(IRGenModule &IGM,
   if (auto theClass = type->getClassOrBoundGenericClass()) {
     // We can access the isas of pure Swift classes directly.
-    if (getRootClass(theClass)->hasKnownSwiftImplementation())
+    if (!theClass->checkAncestry().contains(AncestryFlags::HasImported))
```
Can you throw a convenience method on `ClassDecl` that does this?  It seems to be repeated.

slavapestov replied:
Yeah, that's a good idea.

jrose-apple commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -810,13 +810,15 @@ FUNCTION(UpdateClassMetadata,
               Int8PtrPtrTy->getPointerTo(),
-              SizeTy->getPointerTo()),
+              SizeTy->getPointerTo(),
+              ObjCResilientClassStubTy->getPointerTo()),
```
Sorry, what makes this not an ABI-breaking change?

slavapestov replied:
Adding a new function parameter is not ABI breaking (as long as you don't depend on its value except when you know the caller passed it), but this whole part of the change is going away and instead the stub will be stored in the class context descriptor.

jrose-apple commented on lib/Demangling/Demangler.cpp:
```diff
@@ -1809,6 +1809,10 @@ NodePointer Demangler::demangleMetatype() {
+    case 's':
+      return createWithPoppedType(Node::Kind::ObjCResilientClassStub);
+    case 't':
+      return createWithPoppedType(Node::Kind::FullObjCResilientClassStub);
```
Don't forget to update ABI/Mangling.rst.

jrose-apple commented on test/PrintAsObjC/resilient-ancestry.swift:
```diff
@@ -0,0 +1,29 @@
+// CHECK-NEXT:   - (nonnull instancetype)init OBJC_DESIGNATED_INITIALIZER;
+// CHECK-NEXT: @end
+
+public class ResilientNSObjectSubclass : ResilientNSObjectOutsideParent {}
```
Can you also include a test case for `SWIFT_RESILIENT_CLASS_NAMED` (i.e. an explicit `@objc` name)?

jrose-apple reviewed (commented):
Quick comments, more to come.

slavapestov committed:
AST: Introduce ClassDecl::isSuperclassOf()

slavapestov committed:
Runtime: Add support for Objective-C resilient class stubs

This adds a new tail-allocated field to class context descriptors storing
a pointer to an Objective-C class stub.

When the stub is present, we use the new _objc_realizeClassFromSwift()
entry point to realize the class instead of calling objc_readClassPair().

This should attach categories to the realized class, if they were emitted
to reference the stub.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -3586,72 +3584,57 @@ bool ClassDecl::inheritsSuperclassInitializers(LazyResolver *resolver) {
-  if (Bits.ClassDecl.ObjCKind)
-    return ObjCClassKind(Bits.ClassDecl.ObjCKind - 1);
+  if (Bits.ClassDecl.Ancestry)
+    return AncestryOptions(Bits.ClassDecl.Ancestry - 1);
```
This only works if you can guarantee that you'll never have all flags set. Can you (1) assert that below, and (2) include that in the doc comments of the mutually exclusive flags?

slavapestov replied:
Hmm, you're right that this doesn't make sense. I'll just use a separate bit for AncestryComputed

slavapestov replied:
Wow, I didn't even catch that.

jrose-apple replied:
…and that's not even the right polarity for the necessary condition, so yeah, maybe let's not rely on this.

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -3536,21 +3532,30 @@ enum class ArtificialMainKind : uint8_t {
+  /// different resilience domain.
+  HasResilientOther = (1<<4),
+
+  /// The class or one of its superclasses is imported.
```
What does "imported" mean? Imported from Objective-C? Imported from another module (of any language)?

…can you rename it to HasClangImported? (or just ClangImported, not sure how much the "Has" is buying you on these.)

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3594,9 +3594,11 @@ ERROR(invalid_nonobjc_extension,none,
 ERROR(objc_in_generic_extension,none,
-      "members of extensions of "
-      "%select{classes from generic context|generic classes}0 "
-      "cannot be declared @objc", (bool))
```
I think the distinction between "generic classes" and "classes defined in generic contexts" was serving a useful purpose, please put it back.

slavapestov replied:
I misread the code and thought it was setting it to true always (in 4.2 we had the restriction that classes with generic ancestry could not have categories attached). However it was different for a nested class inside a generic type, since isGeneric() is false in that case.

jrose-apple replied:
Yeah, that's the intent—just showing we understand the code when talking to the user.

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3617,6 +3619,10 @@ NOTE(objc_inference_swift3_addnonobjc,none,
+ERROR(objc_for_resilient_class,none,
+      "classes built with library evolution support cannot have explicit "
+      "'@objc' subclasses because they are not directly "
+      "visible from Objective-C", ())
```
As discussed yesterday, this error message doesn't communicate the right thing to the subclass author who finds themself in this situation.

jrose-apple replied:
We need a diagnostic for the category case too, right?

slavapestov replied:
There is one already

jrose-apple commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -1016,6 +1030,20 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(const ClassDecl *CD) {
+        !ctx.LangOpts.EnableObjCResilientClassStubs) {
+      if (attr->hasName()) {
+        const_cast<ClassDecl *>(CD)->getAttrs().add(
+          new (ctx) ObjCRuntimeNameAttr(*attr));
```
Do we have a test for this behavior already?

slavapestov replied:
I'll try to add one

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -430,7 +430,24 @@ IRGenModule::IRGenModule(IRGenerator &irgen,
+  llvm::Type *params[] = { ObjCClassPtrTy, Int8PtrTy };
+  ObjCUpdateCallbackTy = llvm::FunctionType::get(ObjCClassPtrTy, params, false);
+
+  // The full class stub structure, including a word before the address point.
```
Why do we need to do this? Is it because we treat it like normal Swift metadata, which has the address point at -1 for classes?

slavapestov replied:
It's to work around a linker issue. The linker gets confused when it sees categories that don't point to "real" Objective-C classes. @mikeash knows the details

mikeash replied:
Short version: the linker tries to attach categories to classes at link time when possible. These are not real classes so that goes badly. This makes the class look like a Swift class (because of the data before the ObjC class pointer) which the linker doesn't try to optimize in this way.

rjmccall replied:
Is that going to be permanent ABI, or can we eliminate it in the future with a cooperative linker?

mikeash replied:
We can eliminate it in the future. I think the linker will be able to start applying the category optimization to Swift classes now that Swift has settled down some (Apple folks see rdar://problem/48850004 for that). If it does this, then it will also need to identify these stubs by some other means, such as checking the faux-isa. At that point, we can ditch the pre-class padding word.

jrose-apple reviewed (commented):
Okay, with these comments everything else looks reasonable to me, but I think you should wait for John's final sign-off on the IRGen parts; I don't feel like I understand it well enough to catch potential issues.

slavapestov committed:
Frontend: Add -enable-resilient-objc-class-stubs command line flag

slavapestov committed:
Add mangling and IRGen LinkEntity for Objective-C resilient class stubs

slavapestov committed:
IRGen: Clean up emitObjCClassSymbol() to improve code reuse

slavapestov committed:
IRGen: Emit Objective-C resilient class stubs if experimental flag is on

Non-generic classes with resilient ancestry do not have statically-emitted
metadata, so we can now emit an Objective-C resilient class stub instead.

Also, when emitting an Objective-C category, reference the class stub if
the class has resilient ancestry; previously this case would hit an assert.

Note that class stubs always start with a zero word, with the address point
pointing immediately after. This works around a linker issue, where the
linker tries to coalesce categories and gets confused upon encountering a
class stub.

slavapestov commented:
Everything is done except for the execution tests. I also added tests in a few places for `@_fixed_layout` classes -- note that they still have resilient _metadata_ so the behavior is the same as resilient classes, but I wanted to ensure that we used the right predicate everywhere.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

rjmccall commented on docs/ABI/Mangling.rst:
```diff
@@ -153,6 +152,10 @@ Globals
 
+  global ::= nominal-type 'MU'           // ObjC metadata update callback function
+  global ::= nominal-type 'Ms'           // ObjC resilient class stub
+  global ::= nominal-type 'Mt'           // Full ObjC resilient class stub (private)
```
Why are these down here instead of up next to the other `M` manglings where somebody looking for an available mangling would probably be looking?

slavapestov replied:
Ok, I'll put it back.

rjmccall commented on include/swift/ABI/Metadata.h:
```diff
@@ -3683,6 +3683,20 @@ struct TargetResilientSuperclass {
+struct TargetObjCResilientClassStub {
+  /// A relative pointer to an Objective-C resilient class stub:
+  /// - isa pointer, always 1
+  /// - an update callback, of type 'Class (*)(Class *, objc_class_stub *)'
```
Is this a description of the pointee or something, and if so is there a reason we can't define this structure?  I assume we just don't currently need to interpret it within the Swift runtime.

At any rate, you should clarify that the pointee structure "belongs" to the ObjC ABI.

This structure's name suggests that it's the stub, but it's actually just a pointer to the stub, and the `getStub()` accessor looks through it.  The accessor behavior seems reasonable; you should probably call this `ObjCResilientClassStubInfo` or something similar that clarifies that it's just an optional part of the class descriptor, and that should also be mentioned in a doc comment.

slavapestov replied:
> I assume we just don't currently need to interpret it within the Swift runtime.

Correct. I'll add a better comment and rename this struct.

rjmccall commented on include/swift/ABI/Metadata.h:
```diff
@@ -3759,7 +3777,9 @@ class TargetClassDescriptor final
 
-    // Maybe add something here that's useful only for resilient types?
+    /// Classes with resilient ancestry instead use this flags word to
+    /// indicate the presence of an Objective-C resilient class stub.
```
It looks like this field should always be present and zero in previously-emitted class descriptors, so this should be ABI-compatible.

Consider just saying "otherwise" instead of spelling out a different condition which an uninitiated reader might not understand is the inverse of "does not have a resilient superclass".

"word" is generally used for a pointer-sized chunk in Swift.  You could just "these flags".

Everything after "to" in this comment is over-specific to your patch.  You can say something like "to do things like indicating..." if you think the example is important to reference.

slavapestov replied:
I'll tweak the comment a bit.

rjmccall commented on include/swift/ABI/Metadata.h:
```diff
@@ -3706,7 +3721,8 @@ class TargetClassDescriptor final
                                   TargetOverrideTableHeader<Runtime>,
-                                  TargetMethodOverrideDescriptor<Runtime>>;
+                                  TargetMethodOverrideDescriptor<Runtime>,
+                                  TargetObjCResilientClassStub<Runtime>>;
```
This probably doesn't have to come last since it's not going to be added when back-deploying, but this is good precedent.

rjmccall reviewed (commented):
IRGen changes looks good.  Pretty much all style requests.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test macOS
